src/coreclr/jit/codegenxarch.cpp

```diff
@@ -101,20 +101,40 @@ void CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     if (!pushReg)
     {
+#ifndef TARGET_X86
         // Non-tail call: we can use any callee trash register that is not
-        // a return register or contain 'this' pointer (keep alive this), since
-        // we are generating GS cookie check after a GT_RETURN block.
-        // Note: On Amd64 System V RDX is an arg register - REG_ARG_2 - as well
-        // as return register for two-register-returned structs.
+        // a return register or contain 'this' pointer (keep alive this), or
+        // a continuation register, since we are generating GS cookie check
+        // after a GT_RETURN block.
+        // ARG_1 is EDX or RSI, depending on platform. Either way it fits our requirements
+        regGSCheck = REG_ARG_1;
+
+#else // TARGET_X86
+      // on x86 REG_ARG_1 could be LNGRET_HI, so we will use REG_ARG_0,
+      // unless that contains 'this' pointer (keep alive this), in which case
+      // we use REG_ARG_1, as it could not be LNGRET_HI in such case.
         if (compiler->lvaKeepAliveAndReportThis() && compiler->lvaGetDesc(compiler->info.compThisArg)->lvIsInReg() &&
             (compiler->lvaGetDesc(compiler->info.compThisArg)->GetRegNum() == REG_ARG_0))
         {
             regGSCheck = REG_ARG_1;
         }
+        else if (compiler->compIsAsync())
+        {
+            // in an async method, we can't use use REG_ARG_0 as that is also REG_ASYNC_CONTINUATION_RET
+            // so we use REG_ARG_1 and push/pop it if we have a multireg (long) return.
+            regGSCheck                        = REG_ARG_1;
```

Comment:
hmm, but can REG_ARG_1 contain `this` for the (keep alive this) case?

Reply:
If that is the case we can have a situation when all registers contain GC pointers.

Perhaps just use REG_ARG_0 in async case and force no-gc while we push/pop it.

Reply:
It would be better to model a register kill in LSRA. Similar to the following code, but for `BBJ_RETURN` blocks after we run out of blocks:
https://github.com/dotnet/runtime/blob/d2ee366c51348edd2d057ee075eb3352a78698aa/src/coreclr/jit/lsrabuild.cpp#L2487-L2491

You would pick a volatile registers in most cases but fall back to a callee save in the rare cases. Then you can freely use that in codegen and LSRA will ensure it does not allocate `this` or the generic context to that register.

Reply:
The worst case is that we are 
- returning an object and 
- keeping `this` alive into epilog and 
- we are in an async method 

,so all 3 scratch registers may contain objects. Something will need to be pushed/popped.

Also, I think `GS check + Async` combination will be rare - it is basically a corner case. 
In this scenario we are couple instructions away from epilog, so no-gc seems like a good enough solution that is very local and has less chances to cause some regression.

Reply:
I think for the sake of backporting what you have now looks good, but we should start modelling this properly for .NET 11.

Reply:
> I do not see a good reason to disable GC when we can model the register use properly instead. Also, I do not see how this fixes the live variable problem in general. LSRA may enregister `this` in any register, so it can pick the one being used for the cookie check. The problem is pre-existing, but the only way to fix it is to model things correctly in LSRA. #111230 was an exact case like that.

After thinking about this more, yes it seems a more general problem than just x86 and a more general solution may be desirable in v11. The current fix should be sufficient for v10 and to unblock initial uses/testing of the feature. 
I will log a follow up issue to revisit this.

Reply:
Logged a follow-up issue - https://github.com/dotnet/runtime/issues/119949